core/src/test/java/org/infinispan/test/Mocks.java

```diff
@@ -208,4 +214,20 @@ private Mocks() { }
          blockingPublisher((Subscriber<? super E> innerSubscriber) -> publisher.subscribe(innerSubscriber, complete), checkPoint).subscribe(s);
       };
    }
+
+   /**
+    * Replaces the given component with a spy and returns it for further mocking as needed. Note the original component
+    * is not retrieved and thus requires retrieving before invoking this method if needed.
+    * @param cache the cache to get the component from
+    * @param componentClass the class of the component to retrieve
+    * @param <C> the component class
+    * @return the spied component which has already been replaced and wired in the cache
+    */
+   public static <C> C replaceComponentWithSpy(Cache<?,?> cache, Class<C> componentClass) {
+      C component = TestingUtil.extractComponent(cache, componentClass);
+      C spiedComponent = spy(component);
```

Comment:
If at all possible, I'd rather create a mock that delegates to the existing component instead of using `spy()`.

The instance created by `spy()` has all the annotations, so after copying the state of the existing component (without any synchronization), the component registry will call `start()` on the new instance, which may break things.

Of course, using `mock()` has the opposite problem: the component registry won't call `stop()` on it, which depending on component may hang the test during shutdown. For this reason I've usually written a custom mock class for the component I'm mocking, but it should be possible to use a generic `AbstractMockComponent` class that delegates injection/start/stop and then use `mock(AbstractMockComponent.class, withSettings.additionalInterfaces(componentClass)` to implement the needed interface.

And after writing all of this I realize it's probably too much for this PR, which is merely moving the method...
I'll put it on my TODO list and maybe get to it at some point :)

Reply:
Sounds good.

Reply:
I personally feel that our components should work fine being stopped and restarted, but that may be just a pipe dream.

Reply:
Maybe we could make components restart properly without removing them from the registry and adding a new instance, but that wouldn't really help with `spy()`, because you're copying a component that has already been started and you're starting it again without calling `stop()` before.

Reply:
Gotcha 👍